jetcd-core/src/main/java/io/etcd/jetcd/ClientConnectionManager.java

```diff
@@ -330,35 +331,84 @@ public void onClose(Status status, Metadata trailers) {
+   * @param <T> Converted Type.
+   * @return a CompletableFuture with type T.
+   */
+  public <S, T> CompletableFuture<T> execute(
+          Callable<ListenableFuture<S>> task,long timeout,TimeUnit unit,
+          Function<S, T> resultConvert) {
+    return execute(task,timeout,unit,resultConvert, Util::isRetryable);
+  }
+
+  /**
+   * execute the task and retry it in case of failure.Waits if necessary for at most the given time for result.
+   *
    * @param task a function that returns a new SourceFuture.
+   * @param timeout the maximum time to waits.
+   * @param unit the time unit of the timeout argument.
    * @param resultConvert a function that converts Type S to Type T.
    * @param doRetry a function that determines the retry condition base on SourceFuture error.
    * @param <S> Source type
    * @param <T> Converted Type.
    * @return a CompletableFuture with type T.
    */
   public <S, T> CompletableFuture<T> execute(
-      Callable<ListenableFuture<S>> task,
+      Callable<ListenableFuture<S>> task, long timeout, TimeUnit unit,
       Function<S, T> resultConvert,
       Predicate<Throwable> doRetry) {
 
-    RetryPolicy<S> retryPolicy = new RetryPolicy<S>()
-        .handleIf(doRetry::test)
-        .onRetriesExceeded(e -> newEtcdException(ErrorCode.ABORTED, "maximum number of auto retries reached"))
-        .withBackoff(builder.retryDelay(), builder.retryMaxDelay(), builder.retryChronoUnit());
+    RetryPolicy<S> retryPolicy = getsRetryPolicy(doRetry);
+
+    return Failsafe.with(retryPolicy).with(executorService)
+        .getAsync(() -> {
+          if (timeout >= 0) {
+            return task.call().get(timeout,unit);
+          } else {
+            return task.call().get();
+          }
+        })
+        .thenApply(resultConvert);
+  }
+
+  /**
+   * execute the task and retry it in case of failure.
+   *
+   * @param task a function that returns a new SourceFuture.
+   * @param resultConvert a function that converts Type S to Type T.
+   * @param doRetry a function that determines the retry condition base on SourceFuture error.
+   * @param <S> Source type
+   * @param <T> Converted Type.
+   * @return a CompletableFuture with type T.
+   */
+  public <S, T> CompletableFuture<T> execute(
+          Callable<ListenableFuture<S>> task,
+          Function<S, T> resultConvert,
+          Predicate<Throwable> doRetry) {
+
+    RetryPolicy<S> retryPolicy = getsRetryPolicy(doRetry);
+
+    return Failsafe.with(retryPolicy).with(executorService)
+            .getAsync(() -> task.call().get())
+            .thenApply(resultConvert);
+  }
+
+  private <S> RetryPolicy<S> getsRetryPolicy(Predicate<Throwable> doRetry) {
+    RetryPolicy<S> retryPolicy = new RetryPolicy<S>().handleIf(doRetry::test).onRetriesExceeded(e -> newEtcdException(
+            ErrorCode.ABORTED, "maximum number of auto retries reached")).withBackoff(builder.retryDelay(), builder
+            .retryMaxDelay(), builder.retryChronoUnit());
```

Comment:
the style here is quite not easy to understand, better to reformat it a little

Reply:
here I just extract  common code, and do not make any logic change